### PR TITLE
ci: smarter tav tests for pull requests

### PR DESCRIPTION
**Pull request by v1v**
fixes https://github.com/elastic/apm-agent-nodejs/issues/378

## Highlights
- If the change list contains any related files to the `instrumentation` folders then it will run the TAVs which match them.
- If the GitHub comment contains `jenkins run the module tests for ([module,]+|ALL)` then it will trigger the TAVs for the modules in the comment or all the TAVs if the `ALL` is in the comment.
- Trigger by comment will run all the stages in the Pipeline (aka, build, test, TAV, Docs and so on...)
- `tests_ci` boolean parameter to enable/disable the tests stage in the pipeline when running the Pipeline from the UI. Default: `true`, aka it will run the tests always.
- `TAV *` will be called:
  - either `TAV Test` if it runs all the TAVs 
  - or `TAV Test Subset` if it runs a subset of TAVs and will also include a description `TAV Test comment-triggered` to highlight in GitHub why it was a subset.

## Test Cases
- https://github.com/elastic/apm-agent-nodejs/pull/1228/commits/6da4911ef26efe4d5312cb952bf901a237ee5a96 caused to run the TAV for `handlebars` and `mysql`.

image

image

- Forcing the `params.Run_As_Master_Branch` caused to run all the TAVs. See [build](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-1228/detail/PR-1228/7/pipeline)

- https://github.com/elastic/apm-agent-nodejs/pull/1228#issuecomment-513311236 caused to run the mysql modules only (to clarify the test stage is disabled in this case only for testing purposes):

image

- https://github.com/elastic/apm-agent-nodejs/pull/1228#issuecomment-514148741 caused to run the TAV Subset

image

- `Run_As_Master_Branch` parameter does trigger all the TAVs even though when the above test scenario was still running, that's expected!

image

### CODE REVIEW

**Comment by v1v** [comment]
jenkins run the tests for the module mysql

**Comment by v1v** [comment]
jenkins run the tests for the module mysql

**Comment by v1v** [comment]
jenkins run the tests for the module mysql

**Comment by v1v** [comment]
jenkins run the tests for the module mysql

**Comment by v1v** [comment]
jenkins run the tests for the module mysql

**Comment by v1v** [comment]
jenkins run the tests for the module mysql

**Comment by v1v** [comment]
jenkins run the tests for the module http

**Comment by Qard** [comment]
Is this still WIP? I see passing tests--and it looks great having the separate status checks, by the way!

**Comment by v1v** [comment]
@Qard , as soon as we agree on https://github.com/elastic/apm-agent-nodejs/issues/378 I'll update this PR to be reviewed, does it sound like a good plan?

**Comment by v1v** [comment]
jenkins run the module tests for mysql

File: Jenkinsfile [hunk 1]
```diff
@@ -24,12 +24,13 @@ pipeline {
     quietPeriod(10)
   }
   triggers {
-    issueCommentTrigger('(?i).*(?:jenkins\\W+)?run\\W+(?:the\\W+)?tests(?:\\W+please)?.*')
+    issueCommentTrigger('(?i).*(?:jenkins\\W+)?run\\W+(?:the\\W+)?(?:module\\W+)?tests(?:\\W+please)?.*')
```
Comment (kuisathaverat) [inline Jenkinsfile]: even do the regexp is complimentary, I'd split it in two if it possible, to know that there are two types of comment trigger

File: Jenkinsfile [hunk 2]
```diff
@@ -115,18 +129,42 @@ pipeline {
...
-          unstash 'source'
-          dir("${BASE_DIR}"){
-            script {
-              def node = readYaml(file: '.ci/.jenkins_tav_nodejs.yml')
-              def tav = readYaml(file: '.ci/.jenkins_tav.yml')
+        deleteDir()
+        unstash 'source'
+        dir("${BASE_DIR}"){
+          script {
+            def ghContextName = 'TAV Test'
```
Comment (kuisathaverat) [inline Jenkinsfile]: I'd extract this to a function to give context about what it does.

File: Jenkinsfile (hunk 1 above)
Comment (v1v) [reply Jenkinsfile]: I'm afraid the `issueCommentTrigger` can be declared only once. I could potentially create two groups but the regex would be too long.
The proposed one supports the current use cases:
- `jennkins run the tests` and `jenkins run tests`
- `jenkins run the module tests for` and `jenkins run module tests for`

In any case, the filtering of running TAV is handled later on in the TAV stage.

what do you think if I leave it as it is? Or do you have another use case in mind?

File: Jenkinsfile (hunk 2 above)
Comment (v1v) [reply Jenkinsfile]: See https://github.com/elastic/apm-agent-nodejs/pull/1228/commits/adb4b08e14059269931f4f9c6806a34ef7f7597d

**Pull request merged**